APCAppCore/APCAppCore/Library/Scoring/APCScoring.m

```diff
@@ -281,37 +314,186 @@ - (instancetype)initWithHealthKitQuantityType:(HKQuantityType *)quantityType
+    }
+    
+    //check they are now even
+    NSAssert(self.dataPoints.count == self.correlatedScoring.dataPoints.count, @"Arrays are not of equal length. dataPoints.count = %li, correlatedScoring.dataPoints.count = %li", self.dataPoints.count, self.correlatedScoring.dataPoints.count);
+    
+    //Arrays are not guaranteed to have non-NSNotFound data beginning at the same index
+    NSUInteger dataPointsIndex = [self.dataPoints indexOfObjectPassingTest:^BOOL(id obj, NSUInteger __unused idx, BOOL *stop) {
+        
+        if (![[(NSDictionary *)obj objectForKey:kDatasetValueKey] isEqualToNumber: @(NSNotFound)]) {
+            *stop = YES;
+            return YES;
+        }else{
+            return NO;
+        }
+    }];
+    
+    NSUInteger correlatedDataPointsIndex = [self.correlatedScoring.dataPoints indexOfObjectPassingTest:^BOOL(id obj, NSUInteger __unused idx, BOOL *stop) {
+        if (![[(NSDictionary *)obj objectForKey:kDatasetValueKey] isEqualToNumber: @(NSNotFound)]) {
+            *stop = YES;
+            return YES;
+        }else{
+            return NO;
+        }
+    }];
+    
+    NSUInteger highestIndex = 0;
+    if (dataPointsIndex >= correlatedDataPointsIndex) {
+        highestIndex = dataPointsIndex;
+    }else if(correlatedDataPointsIndex > dataPointsIndex){
+        highestIndex = correlatedDataPointsIndex;
+    }
+    
+    if (correlatedDataPointsIndex != NSNotFound && dataPointsIndex != NSNotFound && self.dataPoints.count == self.correlatedScoring.dataPoints.count){
+        for (NSUInteger i = 0; i < highestIndex; i++) {
+            NSMutableDictionary *dataPointDictionary = [[self.dataPoints objectAtIndex:i] mutableCopy];
+            [dataPointDictionary setObject:@(NSNotFound) forKey:kDatasetValueKey];
+            [self.dataPoints replaceObjectAtIndex:i withObject:dataPointDictionary];
+            
+            NSMutableDictionary *correlatedDataPointDictionary =[[self.correlatedScoring.dataPoints objectAtIndex:i] mutableCopy];
+            [correlatedDataPointDictionary setObject:@(NSNotFound) forKey:kDatasetValueKey];
+            [self.correlatedScoring.dataPoints replaceObjectAtIndex:i withObject:correlatedDataPointDictionary];
+            
+        }
+    }
+}
+
+-(void)indexDataSeries:(NSMutableArray *)series
+{
+    
+    NSDictionary *basePointObject;
+    NSNumber *basePointValue = @0;
+    
+    //find the earliest base point value
+    for (int i = (int)series.count -1; i >= 0; i--) {
+        basePointObject = series[i];
+        NSNumber *checkBasePointValue = [basePointObject valueForKey:kDatasetValueKey];
+        if (![checkBasePointValue  isEqual: @(NSNotFound)]) {
+            basePointValue = checkBasePointValue;
+        }
+    }
+    
+    //loop over all elements calculating the point index
+    NSNumber *index;
+    for (NSUInteger i = 0; i < series.count; i++) {
+        NSMutableDictionary *dictionary = [[series objectAtIndex:i] mutableCopy];
```

Comment:
Why are we making a mutable copy here? Better to make the copy inside the conditional so we're not needlessly copying any NSNotFound records.